Add IMX519 driver and device tree overlay

ArduCAM opened the pull request.

Commit:
configs: Add CONFIG_VIDEO_IMX519=m
Include the driver module for the IMX519.

Signed-off-by: Lee Jackson <info@arducam.com>

6by9 commented on arch/arm/boot/dts/overlays/imx519-overlay.dts:
```diff
@@ -0,0 +1,113 @@
+	};
+
+	__overrides__ {
+		rotation = <&imx519>,"rotation:0";
```
Could you add the orientation parameter as done in https://github.com/raspberrypi/linux/commit/c3c2ca3c775644baf80dde6a9a79f745b18ce108 please?

6by9 commented on arch/arm/boot/dts/overlays/imx519-overlay.dts:
```diff
@@ -0,0 +1,113 @@
+// SPDX-License-Identifier: GPL-2.0-only
```
Please add an entry in README for the new overlay.

6by9 commented on drivers/media/i2c/Makefile:
```diff
@@ -128,3 +128,4 @@ obj-$(CONFIG_VIDEO_RDACM20)	+= rdacm20-camera_module.o
 obj-$(CONFIG_VIDEO_ST_MIPID02) += st-mipid02.o
 
 obj-$(CONFIG_SDR_MAX2175) += max2175.o
+obj-$(CONFIG_VIDEO_IMX519) += imx519.o
```
Whilst not in alphabetical order (and actually it's our downstream imx477 driver that is out of place), the drivers are at least grouped reasonably.
Please add your driver after IMX355.

6by9 commented on drivers/media/i2c/imx519.c:
```diff
@@ -0,0 +1,2321 @@
+#define IMX519_FRAME_LENGTH_MAX		0xffdc
+
+/* Long exposure multiplier */
+#define IMX519_LONG_EXP_SHIFT_MAX	7
```
I don't have a datasheet for IMX519. Does it support a long exposure multiplier register in the same way as IMX477?

ArduCAM replied:
Yes, it is the same as imx477.

6by9 commented on drivers/media/i2c/imx519.c:
```diff
@@ -0,0 +1,2321 @@
+static const struct imx519_reg mode_common_regs[] = {
+	{0x0136, 0x18},
+	{0x0137, 0x00},
+	{0x3C7E, 0x01},
```
The normal coding style is to use lower case for hex values. We have a bit of a mixture in this driver at present.

6by9 commented on drivers/media/i2c/imx519.c:
```diff
@@ -0,0 +1,2321 @@
+		.height = 2160,
+		.line_length_pix = 0x3870,
+		.crop = {
+			.left = IMX519_PIXEL_ARRAY_LEFT + 0x0198,
```
Could we use decimal here please? I realise this comes from a register that is programmed in hex (presumably 0x0344/5), but it just makes it more readable. Ditto the other similar offsets.

6by9 commented on drivers/media/i2c/imx519.c:
```diff
@@ -0,0 +1,2321 @@
+#define IMX519_XCLR_MIN_DELAY_US	8000
+#define IMX519_XCLR_DELAY_RANGE_US	1000
+
+struct imx519_compatible_data {
```
Not needed for this driver as there is only one compatible string.

6by9 commented on drivers/media/i2c/imx519.c:
```diff
@@ -0,0 +1,2321 @@
+ * in the datasheet) is much smaller - 600us.
+ */
+#define IMX519_XCLR_MIN_DELAY_US	8000
+#define IMX519_XCLR_DELAY_RANGE_US	1000
```
This is a straight copy from IMX477. Does it actually apply to IMX519?

ArduCAM replied:
From the data sheet, they are the same.

6by9 commented on drivers/media/i2c/imx519.c:
```diff
@@ -0,0 +1,2321 @@
+	unsigned int fmt_code;
+
+	struct clk *xclk;
+	u32 xclk_freq;
```
I realise this is copied from imx477, but xclk_freq is only accessed in imx477_probe, so can be a local variable there. There's currently no expectation to need to work off multiple clock frequencies.

6by9 commented on drivers/media/i2c/imx519.c:
```diff
@@ -0,0 +1,2321 @@
+	return container_of(_sd, struct imx519, sd);
+}
+
+static inline void get_mode_table(unsigned int code,
```
Not needed as we only have the one table, so it can be accessed directly.
(IMX477 has a different list of modes for 10 and 12 bit).

6by9 commented on drivers/media/i2c/imx519.c:
```diff
@@ -0,0 +1,2321 @@
+	if (i >= ARRAY_SIZE(codes))
+		i = 0;
+
+	i = (i & ~3) | (imx519->vflip->val ? 2 : 0) |
```
```(i & ~3)``` is redundant as this only supports 10 bit modes.

6by9 commented on drivers/media/i2c/imx519.c:
```diff
@@ -0,0 +1,2321 @@
+
+
+/* Mode configs */
+static const struct imx519_mode supported_modes_10bit[] = {
```
Does the sensor support 8 or 12 bit readout, and do they work? If not then this can be called just ```supported_modes```.

ArduCAM replied:
imx519 supports RAW8, but I haven't added RAW8 config yet.

6by9 replied:
Is there a sensible use case for RAW8? Can it actually run faster or similar? If not then I wouldn't worry about it.

You may get others adding support at a later date (as happened for imx219), but that doesn't need to be your concern initially.

6by9 commented on drivers/media/i2c/imx519.c:
```diff
@@ -0,0 +1,2321 @@
+
+#define IMX519_XCLK_FREQ		24000000
+
+#define IMX519_DEFAULT_LINK_FREQ	648000000
```
Minor note that if this is actually correct, then it exceeds the spec for CSI-2 which is 1Gbit/s/lane, which equates to 500MHz as it is DDR.
Unicam (the CSI-2 receiver) does normally cope with slightly higher data rates, but technically you're exceeding the spec.

ArduCAM replied:
This should be my calculation error. 
The actual frequency was 525Mhz before, but it has been reduced to 493.5Mhz.

6by9 reviewed (commented):
A number of these look to be due to copying IMX477 but not supporting some of the features. This leaves a number of chunks of effectively dead code around.

Commit:
media: i2c: Add driver for IMX519 sensor
Adds a driver for the 16MPix IMX519 CSI2 sensor.
Whilst the sensor supports 2 or 4 CSI2 data lanes, this driver
currently only supports 2 lanes.

The following Bayer modes are currently available:

4656x3496 10-bit @ 10fps
3840x2160 10-bit (cropped) @ 21fps
2328x1748 10-bit (binned) @ 30fps
1920x1080 10-bit (cropped/binned) @ 60fps
1280x720 10-bit (cropped/binned) @ 120fps

Signed-off-by: Lee Jackson <info@arducam.com>

Commit:
media: dt-bindings: media: i2c: Add IMX519 CMOS sensor binding
Add YAML device tree binding for IMX519 CMOS image sensor, and
the relevant MAINTAINERS entries.

Signed-off-by: Lee Jackson <info@arducam.com>

Commit:
overlays: Add imx519-overlay.dts
Added overlays for enabling IMX519 and add the
relevant information to the README.

Signed-off-by: Lee Jackson <info@arducam.com>

6by9 commented on drivers/media/i2c/imx519.c:
```diff
@@ -0,0 +1,2107 @@
+
+	lockdep_assert_held(&imx519->mutex);
+
+	for (i = 0; i < ARRAY_SIZE(codes); i++)
+		if (codes[i] == code)
+			break;
+
+	if (i >= ARRAY_SIZE(codes))
+		i = 0;
+
```
We don't need this bit either as it's also still related to supporting multiple bit depths.

ArduCAM replied:
Maybe I can keep this? Modifying the code here needs to change a way to obtain Bayer order, and also to adapt to various calls.

6by9 replied:
You're computing a value of ```i``` which is then not used.
If you add support for 8 bit readout from the start then you need to retain this, but otherwise it's redundant. If sent upstream then you'd get a very similar comment as it serves no purpose.

It does make ```code``` unused too, so that could be removed too.

ArduCAM replied:
Sorry, I confuse the method of calculating the Bayer order here with the method used elsewhere.
The `i` here is indeed not used.

6by9 commented on drivers/media/i2c/imx519.c:
```diff
@@ -0,0 +1,2107 @@
+	/* Set default mode to max resolution */
+	imx519->mode = &supported_modes_10bit[0];
+	imx519->fmt_code = MEDIA_BUS_FMT_SRGGB10_1X10;
+}
```
This is only called from one place in imx519_probe, so may as well just be inserted there.

6by9 reviewed (commented):
DT bindings needs to come before the driver. MAINTAINERS normally gets updated once with the bindings to add both bindings and the driver files, eg https://github.com/torvalds/linux/commit/8a97a4676f8b1badcd9cfbed2b081342847bb1b1

By copying imx477 I'm afraid you've got caught in the cross-fire slightly between us and mainline over supporting multiple streams (ie image with embedded metadata). Mainline still haven't resolved this, although patches have been kicked around recently. We adopted just advertising 2 pads for the two streams, but that isn't be acceptable to upstream.

If you're looking to upstream this, then the references to METADATA_PAD need to be removed, and all the pad index checks become ```if (thing->pad)``` instead of ```if (things->pad >= NUM_PADS)```. I'm happy to accept an additional patch to our tree which adds the second pad back in, and proper support for the streams can be added when they've decided how to do it.

ArduCAM commented:
Hi 6by9
Thank you very much for your review and suggestions,
I realize that I have made a lot of stupid mistakes that have caused you unnecessary trouble, and I am sorry for that.

6by9 commented:
No problems. I'm happy to review drivers before submission upstream as I recognise that it is a fairly steep learning curve.
Interactions on the linux-media mailing list can be slow and tedious, and often disheartening, with a number of seemingly unwritten rules that you have to pick up over time.

My comment still applies over the METADATA_PAD - that isn't the way upstream are going to want it, so simplest to implement just exposing video initially. Add the metadata pad later on if we have code that will parse it.
I need to investigate the upstream multiplexed stream support and see how it is going to work - https://patchwork.linuxtv.org/project/linux-media/list/?series=6197 / https://git.kernel.org/pub/scm/linux/kernel/git/tomba/linux.git/log/?h=multistream/work-v8

ArduCAM commented:
Yes, I saw your description of METADATA_PAD, I was thinking maybe we have a chance to merge it into the Raspberry Pi kernel first?

ArduCAM commented:
Hi @6by9 
If we want to merge into the kernel of raspberry pi first, do we need to delete MATEDATA_PAD?

6by9 commented:
Sorry, I'd missed your previous comment.

We try to minimise any differences to mainline, unless a feature or change is totally unacceptable to mainline (eg the abuse of pads for the metadata stream).
Even those changes should be made as an extra patch on top of what should be acceptable to mainline.

So it really would be preferable to have the METADATA_PAD change split out as a separate patch. You need to do that for mainline anyway.

ArduCAM committed:
media: i2c: imx519: Advertise embedded data node on media pad 1
This commit updates the imx519 driver to adverise support for embedded
data streams.

The imx519 sensor subdevice overloads the media pad to differentiate
between image stream (pad 0) and embedded data stream (pad 1) when
performing the v4l2_subdev_pad_ops functions.

Signed-off-by: Lee Jackson <info@arducam.com>

ArduCAM commented:
Hi @6by9 
Thanks for your suggestion, do you think this patch is okay?

ArduCAM commented:
Hi @6by9 
_Just in case you missed this._

6by9 commented:
Apologies, I had missed the update again.

That looks good now. There may be a couple of quibbles upstream over minor details of the implementation, but most of it looks fine to me.
I'm trusting that it actually produces images and the controls do what they're supposed to.

pelwell commented:
These patches required some adjustments for the newer kernels, but they're in rpi-5.14.y and rpi-5.15.y now (along with blank lines after the first lines of the commit messages).

ArduCAM commented:
Got it, I will test it on the new version of the kernel.

ArduCAM commented:
Hi @pelwell 
After testing everything looks fine.
![2021-10-18-103352_1920x1080_scrot](https://user-images.githubusercontent.com/11016968/137832968-18f8ae1d-7848-4928-bbbd-7cbd48d89ac0.png)
![2021-10-19-031608_1920x1080_scrot](https://user-images.githubusercontent.com/11016968/137833001-9e207cca-a3e1-4c33-b63b-b4c9471ca0fd.png)
